src/Bicep.LangServer/Handlers/BicepDecompileSaveCommandHandler.cs

```diff
@@ -0,0 +1,274 @@
+        )
+        {
+            StringBuilder output = new StringBuilder();
+
+            try
+            {
+                if (outputFiles.Length == 0)
+                {
+                    throw new ArgumentException($"{nameof(outputFiles)} should not be empty");
+                }
+
+                string proposedMainBicepPath = outputFiles[0].absolutePath;
+                string? outputFolder = Path.GetDirectoryName(proposedMainBicepPath);
+                if (outputFolder is null)
+                {
+                    throw new ArgumentException($"Invalid input path {proposedMainBicepPath}");
+                }
+
+                // Figure out proper place to save files
+                (string path, string contents)[] filesToSave = DeterminePathsToSave(output, ref outputFolder, proposedMainBicepPath, outputFiles, overwrite);
+                if (!overwrite)
+                {
+                    Debug.Assert(filesToSave.All(f => IsPathRelativeToBasePath(f.path, outputFolder)), $"Expected all output files to be relative to {outputFolder}");
+                }
+                var actualMainBicepPath = filesToSave[0].path;
+
+                // Save files
+                SaveFiles(output, filesToSave);
+
+                // Completion message
+                Log(output, $"Decompilation complete.");
+
+                // Show main output file
+                if (this.clientCapabilitiesProvider.DoesClientSupportShowDocumentRequest())
+                {
+                    await this.languageServerFacade.Window.ShowDocument(
+                        new() { TakeFocus = true, Uri = actualMainBicepPath },
+                        CancellationToken.None);
+                }
+
+                // Return result
+                return (
+                    result: new BicepDecompileSaveCommandResult(
+                        output.ToString(),
+                        null,
+                        actualMainBicepPath,
+                        filesToSave.Select(f => f.path).ToArray()),
+                    successTelemetry: BicepTelemetryEvent.DecompileSaveSuccess(decompileId)
+                    );
+            }
+            catch (Exception ex)
+            {
+                Log(output, ex.Message);
+                throw telemetryHelper.CreateException(
+                    ex.Message,
+                    BicepTelemetryEvent.DecompileSaveFailure(decompileId, ex.GetType().Name),
+                    new BicepDecompileSaveCommandResult(
+                        output.ToString(),
+                        ex.Message,
+                        null,
+                        new string[] { })
+                );
+            }
+        }
+
+        private (string path, string contents)[] DeterminePathsToSave(StringBuilder output, ref string outputFolder, string mainBicepPath, DecompiledFile[] outputFiles, bool overwrite)
+        {
+            var singleFileDecompilation = outputFiles.Length == 1;
```

Comment:
If user chose overwrite, then this is useless. Can we please move it after if (overwrite) {..} check?

Reply:
[sw/decompile2 d9b3c902b] CR fixes